core/arch/x86/memfuncs.asm

```diff
@@ -0,0 +1,132 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Copyright (c) 2003-2007 Determina Corp. */
+/* Copyright (c) 2001-2003 Massachusetts Institute of Technology */
+/* Copyright (c) 2001 Hewlett-Packard Company */
+
+/*
+ * memfuncs.asm: Contains our custom memcpy and memset routines.
+ *
+ * The core certainly needs its own memcpy and memset for libc isolation as
+ * described below, when a shared library and used for early injection.
+ * We also use these to avoid glibc-versioned imports in our auxiliary tools
+ * such as drrun (i#1504).  However, when DR is built as a static library, these
+ * functions can conflict with the application's own versions (even when
+ * marking as weak): i#3348.  Since a static-library DR will not be taking control
+ * early, and since these functions should always be re-entrant, we simply
+ * do not include them in libdynamorio_static and it imports from libc.
+ * We accomplish that by separating these into their own library.
+ */
+
+#include "../asm_defines.asm"
+#include "x86_asm_defines.asm" /* PUSHGPR, POPGPR, etc. */
+START_FILE
+
+/***************************************************************************/
+#ifdef UNIX
+
+/* i#46: Implement private memcpy and memset for libc isolation.  If we import
+ * memcpy and memset from libc in the normal way, the application can override
+ * those definitions and intercept them.  In particular, this occurs when
+ * running an app that links in the Address Sanitizer runtime.  Since we already
+ * need a reasonably efficient assembly memcpy implementation for safe_read, we
+ * go ahead and reuse the code for private memcpy and memset.
```

Comment:
I don't see an explicit reference to safe_read in here, but im guessing the comment is referring to these macros which might be shared? Do you mind clearing up exactly what parts or shared, or fixing this comment up if the following functions don't actually reuse stuff which is specifically for safe_read?

Reply:
safe_read_asm shares the REP_STRING_OP macro which is also used here.

This comment does bring up something I did not make clear in the issue: there is a downside to having static DR use glibc memcpy.  It lets sanitizers or other interceptors override
DR's memcpy.  However, we already have interoperability concerns with sanitizers, so I was assuming we could live with that.

Reply:
I want to also add the sanitizer concern to the code comments and not just the issue so I can send you a comment update PR.

Reply:
PR #3376